diff --git a/consensus/reactor_test.go b/consensus/reactor_test.go
index d5dae78dc..7b9381898 100644
--- a/consensus/reactor_test.go
+++ b/consensus/reactor_test.go
@@ -31,6 +31,7 @@ import (
 	"github.com/tendermint/tendermint/libs/log"
 	cmtsync "github.com/tendermint/tendermint/libs/sync"
 	mempl "github.com/tendermint/tendermint/mempool"
+	mempoolv2 "github.com/tendermint/tendermint/mempool/cat"
 	mempoolv0 "github.com/tendermint/tendermint/mempool/v0"
 	mempoolv1 "github.com/tendermint/tendermint/mempool/v1"
 	"github.com/tendermint/tendermint/p2p"
@@ -185,6 +186,16 @@ func TestReactorWithEvidence(t *testing.T) {
 				mempoolv1.WithPreCheck(sm.TxPreCheck(state)),
 				mempoolv1.WithPostCheck(sm.TxPostCheck(state)),
 			)
+		case cfg.MempoolV2:
+			mempool = mempoolv2.NewTxPool(
+				logger,
+				config.Mempool,
+				proxyAppConnConMem,
+				state.LastBlockHeight,
+				mempoolv2.WithMetrics(memplMetrics),
+				mempoolv2.WithPreCheck(sm.TxPreCheck(state)),
+				mempoolv2.WithPostCheck(sm.TxPostCheck(state)),
+			)
 		}
 		if thisConfig.Consensus.WaitForTxs() {
 			mempool.EnableTxsAvailable()
@@ -698,7 +709,7 @@ func timeoutWaitGroup(t *testing.T, n int, f func(int), css []*State) {
 		close(done)
 	}()
 
-	// we're running many nodes in-process, possibly in in a virtual machine,
+	// we're running many nodes in-process, possibly in a virtual machine,
 	// and spewing debug messages - making a block could take a while,
 	timeout := time.Second * 120
 
