diff --git a/test/e2e/generator/random.go b/test/e2e/generator/random.go
index ec59a01b2..75e5fc5a2 100644
--- a/test/e2e/generator/random.go
+++ b/test/e2e/generator/random.go
@@ -63,12 +63,14 @@ func (wc weightedChoice) Choose(r *rand.Rand) interface{} {
 	total := 0
 	choices := make([]interface{}, 0, len(wc))
 	for choice, weight := range wc {
+		//nolint:gosec
 		total += int(weight)
 		choices = append(choices, choice)
 	}
 
 	rem := r.Intn(total)
 	for _, choice := range choices {
+		//nolint:gosec
 		rem -= int(wc[choice])
 		if rem <= 0 {
 			return choice
