diff --git a/p2p/conn/secret_connection.go b/p2p/conn/secret_connection.go
index 89adf06f7..4201b153e 100644
--- a/p2p/conn/secret_connection.go
+++ b/p2p/conn/secret_connection.go
@@ -207,6 +207,7 @@ func (sc *SecretConnection) Write(data []byte) (n int, err error) {
 				data = nil
 			}
 			chunkLength := len(chunk)
+			//nolint:gosec
 			binary.LittleEndian.PutUint32(frame, uint32(chunkLength))
 			copy(frame[dataLenSize:], chunk)
 
