diff --git a/rpc/test/helpers.go b/rpc/test/helpers.go
index 34dc19a2d..8cc9a1d34 100644
--- a/rpc/test/helpers.go
+++ b/rpc/test/helpers.go
@@ -27,6 +27,8 @@ import (
 type Options struct {
 	suppressStdout bool
 	recreateConfig bool
+	// SpecificConfig will replace the global config if not nil
+	SpecificConfig *cfg.Config
 }
 
 var globalConfig *cfg.Config
@@ -115,6 +117,15 @@ func GetGRPCClient() core_grpc.BroadcastAPIClient {
 	return core_grpc.StartGRPCClient(grpcAddr)
 }
 
+func GetBlockAPIClient() (core_grpc.BlockAPIClient, error) {
+	grpcAddr := globalConfig.RPC.GRPCListenAddress
+	client, err := core_grpc.StartBlockAPIGRPCClient(grpcAddr)
+	if err != nil {
+		return nil, err
+	}
+	return client, nil
+}
+
 // StartTendermint starts a test CometBFT server in a go routine and returns when it is initialized
 func StartTendermint(app abci.Application, opts ...func(*Options)) *nm.Node {
 	nodeOpts := defaultOptions
@@ -151,7 +162,11 @@ func StopTendermint(node *nm.Node) {
 // NewTendermint creates a new CometBFT server and sleeps forever
 func NewTendermint(app abci.Application, opts *Options) *nm.Node {
 	// Create & start node
+	if opts.SpecificConfig != nil {
+		globalConfig = opts.SpecificConfig
+	}
 	config := GetConfig(opts.recreateConfig)
+
 	var logger log.Logger
 	if opts.suppressStdout {
 		logger = log.NewNopLogger()
