diff --git a/p2p/metrics.go b/p2p/metrics.go
index 7c80658e5..9da34d47c 100644
--- a/p2p/metrics.go
+++ b/p2p/metrics.go
@@ -87,13 +87,13 @@ func PrometheusMetrics(namespace string, labelsAndValues ...string) *Metrics {
 			Subsystem: MetricsSubsystem,
 			Name:      "message_receive_bytes_total",
 			Help:      "Number of bytes of each message type received.",
-		}, append(labels, "message_type")).With(labelsAndValues...),
+		}, append(labels, "message_type", "chID", "peer_id")).With(labelsAndValues...),
 		MessageSendBytesTotal: prometheus.NewCounterFrom(stdprometheus.CounterOpts{
 			Namespace: namespace,
 			Subsystem: MetricsSubsystem,
 			Name:      "message_send_bytes_total",
 			Help:      "Number of bytes of each message type sent.",
-		}, append(labels, "message_type")).With(labelsAndValues...),
+		}, append(labels, "message_type", "chID", "peer_id")).With(labelsAndValues...),
 	}
 }
 
