diff --git a/p2p/netaddress.go b/p2p/netaddress.go
index b8e0c2419..dec331758 100644
--- a/p2p/netaddress.go
+++ b/p2p/netaddress.go
@@ -57,6 +57,7 @@ func NewNetAddress(id ID, addr net.Addr) *NetAddress {
 	}
 
 	ip := tcpAddr.IP
+	//nolint:gosec
 	port := uint16(tcpAddr.Port)
 	na := NewNetAddressIPPort(ip, port)
 	na.ID = id
@@ -146,8 +147,9 @@ func NetAddressFromProto(pb tmp2p.NetAddress) (*NetAddress, error) {
 		return nil, fmt.Errorf("invalid port number %v", pb.Port)
 	}
 	return &NetAddress{
-		ID:   ID(pb.ID),
-		IP:   ip,
+		ID: ID(pb.ID),
+		IP: ip,
+		//nolint:gosec
 		Port: uint16(pb.Port),
 	}, nil
 }
