diff --git a/p2p/peer_set_test.go b/p2p/peer_set_test.go
index 6501dd77a..ca92c65cb 100644
--- a/p2p/peer_set_test.go
+++ b/p2p/peer_set_test.go
@@ -27,6 +27,7 @@ func (mp *mockPeer) NodeInfo() NodeInfo              { return DefaultNodeInfo{}
 func (mp *mockPeer) Status() ConnectionStatus        { return ConnectionStatus{} }
 func (mp *mockPeer) ID() ID                          { return mp.id }
 func (mp *mockPeer) IsOutbound() bool                { return false }
+func (mp *mockPeer) HasIPChanged() bool              { return false }
 func (mp *mockPeer) IsPersistent() bool              { return true }
 func (mp *mockPeer) Get(s string) interface{}        { return s }
 func (mp *mockPeer) Set(string, interface{})         {}
