diff --git a/p2p/types.go b/p2p/types.go
index 7a741bd29..427f58035 100644
--- a/p2p/types.go
+++ b/p2p/types.go
@@ -16,6 +16,12 @@ type Envelope struct {
 	ChannelID byte
 }
 
+type UnprocessedEnvelope struct {
+	Src       IntrospectivePeer
+	Message   []byte
+	ChannelID byte
+}
+
 // Unwrapper is a Protobuf message that can contain a variety of inner messages
 // (e.g. via oneof fields). If a Channel's message type implements Unwrapper, the
 // p2p layer will automatically unwrap inbound messages so that reactors do not have to do this themselves.
