diff --git a/cmd/cometbft/commands/debug/io.go b/cmd/cometbft/commands/debug/io.go
index ec8e04e0e..444bf4fe6 100644
--- a/cmd/cometbft/commands/debug/io.go
+++ b/cmd/cometbft/commands/debug/io.go
@@ -109,5 +109,5 @@ func writeStateJSONToFile(state interface{}, dir, filename string) error {
 		return fmt.Errorf("failed to encode state dump: %w", err)
 	}
 
-	return os.WriteFile(path.Join(dir, filename), stateJSON, os.ModePerm)
+	return os.WriteFile(path.Join(dir, filename), stateJSON, os.ModePerm) //nolint:gosec
 }
