diff --git a/cmd/cometbft/commands/debug/util.go b/cmd/cometbft/commands/debug/util.go
index d471eb959..ff777b9bb 100644
--- a/cmd/cometbft/commands/debug/util.go
+++ b/cmd/cometbft/commands/debug/util.go
@@ -79,5 +79,5 @@ func dumpProfile(dir, addr, profile string, debug int) error {
 		return fmt.Errorf("failed to read %s profile response body: %w", profile, err)
 	}
 
-	return os.WriteFile(path.Join(dir, fmt.Sprintf("%s.out", profile)), body, os.ModePerm)
+	return os.WriteFile(path.Join(dir, fmt.Sprintf("%s.out", profile)), body, os.ModePerm) //nolint:gosec
 }
