diff --git a/test/maverick/consensus/wal.go b/test/maverick/consensus/wal.go
index 2c7ccc2a2..adfc24fc0 100644
--- a/test/maverick/consensus/wal.go
+++ b/test/maverick/consensus/wal.go
@@ -177,7 +177,7 @@ func (wal *BaseWAL) WriteSync(msg cmtcon.WALMessage) error {
 	}
 
 	if err := wal.FlushAndSync(); err != nil {
-		wal.Logger.Error(`WriteSync failed to flush consensus wal. 
+		wal.Logger.Error(`WriteSync failed to flush consensus wal.
 		WARNING: may result in creating alternative proposals / votes for the current height iff the node restarted`,
 			"err", err)
 		return err
@@ -284,6 +284,7 @@ func (enc *WALEncoder) Encode(v *cmtcon.TimedWALMessage) error {
 	}
 
 	crc := crc32.Checksum(data, crc32c)
+	//nolint:gosec
 	length := uint32(len(data))
 	if length > maxMsgSizeBytes {
 		return fmt.Errorf("msg is too big: %d bytes, max: %d bytes", length, maxMsgSizeBytes)
