diff --git a/blockchain/v2/processor.go b/blockchain/v2/processor.go
index 8b7401e7d..3d5097e80 100644
--- a/blockchain/v2/processor.go
+++ b/blockchain/v2/processor.go
@@ -178,7 +178,7 @@ func (state *pcState) handle(event Event) (Event, error) {
 
 		state.context.saveBlock(first, firstParts, second.LastCommit)
 
-		if err := state.context.applyBlock(firstID, first); err != nil {
+		if err := state.context.applyBlock(firstID, first, second.LastCommit); err != nil {
 			panic(fmt.Sprintf("failed to process committed block (%d:%X): %v", first.Height, first.Hash(), err))
 		}
 
