diff --git a/blockchain/v2/processor_context.go b/blockchain/v2/processor_context.go
index fed488d09..d09a89770 100644
--- a/blockchain/v2/processor_context.go
+++ b/blockchain/v2/processor_context.go
@@ -8,7 +8,7 @@ import (
 )
 
 type processorContext interface {
-	applyBlock(blockID types.BlockID, block *types.Block) error
+	applyBlock(blockID types.BlockID, block *types.Block, seenCommit *types.Commit) error
 	verifyCommit(chainID string, blockID types.BlockID, height int64, commit *types.Commit) error
 	saveBlock(block *types.Block, blockParts *types.PartSet, seenCommit *types.Commit)
 	cmtState() state.State
@@ -29,8 +29,8 @@ func newProcessorContext(st blockStore, ex blockApplier, s state.State) *pContex
 	}
 }
 
-func (pc *pContext) applyBlock(blockID types.BlockID, block *types.Block) error {
-	newState, _, err := pc.applier.ApplyBlock(pc.state, blockID, block)
+func (pc *pContext) applyBlock(blockID types.BlockID, block *types.Block, seenCommit *types.Commit) error {
+	newState, _, err := pc.applier.ApplyBlock(pc.state, blockID, block, seenCommit)
 	pc.state = newState
 	return err
 }
@@ -68,7 +68,7 @@ func newMockProcessorContext(
 	}
 }
 
-func (mpc *mockPContext) applyBlock(blockID types.BlockID, block *types.Block) error {
+func (mpc *mockPContext) applyBlock(blockID types.BlockID, block *types.Block, seenCommit *types.Commit) error {
 	for _, h := range mpc.applicationBL {
 		if h == block.Height {
 			return fmt.Errorf("generic application error")
