diff --git a/blockchain/v2/reactor.go b/blockchain/v2/reactor.go
index a381572c5..7c3330a52 100644
--- a/blockchain/v2/reactor.go
+++ b/blockchain/v2/reactor.go
@@ -55,7 +55,7 @@ type blockVerifier interface {
 }
 
 type blockApplier interface {
-	ApplyBlock(state state.State, blockID types.BlockID, block *types.Block) (state.State, int64, error)
+	ApplyBlock(state state.State, blockID types.BlockID, block *types.Block, seenCommit *types.Commit) (state.State, int64, error)
 }
 
 // XXX: unify naming in this package around cmtState
