diff --git a/scripts/mockery_generate.sh b/scripts/mockery_generate.sh
index 8b97f91e2..e03c86a9c 100755
--- a/scripts/mockery_generate.sh
+++ b/scripts/mockery_generate.sh
@@ -3,5 +3,4 @@
 # Invoke Mockery v2 to update generated mocks for the given type.
 # Last change was made based on changes for main in https://github.com/tendermint/tendermint/pull/9196 
 
-
-go run github.com/vektra/mockery/v2 --disable-version-string --case underscore --name "$*"
+go run github.com/vektra/mockery/v2@latest --disable-version-string --case underscore --name "$*"
