diff --git a/p2p/pex/addrbook.go b/p2p/pex/addrbook.go
index a6a6c6dbc..63218692d 100644
--- a/p2p/pex/addrbook.go
+++ b/p2p/pex/addrbook.go
@@ -681,6 +681,7 @@ func (a *addrBook) addAddress(addr, src *p2p.NetAddress) error {
 			return nil
 		}
 		// The more entries we have, the less likely we are to add more.
+		//nolint:gosec
 		factor := int32(2 * len(ka.Buckets))
 		if a.rand.Int31n(factor) != 0 {
 			return nil
@@ -849,6 +850,7 @@ func (a *addrBook) calcNewBucket(addr, src *p2p.NetAddress) (int, error) {
 	if err != nil {
 		return 0, err
 	}
+	//nolint:gosec
 	result := int(binary.BigEndian.Uint64(hash2) % newBucketCount)
 	return result, nil
 }
@@ -875,6 +877,7 @@ func (a *addrBook) calcOldBucket(addr *p2p.NetAddress) (int, error) {
 	if err != nil {
 		return 0, err
 	}
+	//nolint:gosec
 	result := int(binary.BigEndian.Uint64(hash2) % oldBucketCount)
 	return result, nil
 }
