diff --git a/behaviour/peer_behaviour.go b/behaviour/peer_behaviour.go
index f7cfd00f0..652e44e73 100644
--- a/behaviour/peer_behaviour.go
+++ b/behaviour/peer_behaviour.go
@@ -25,7 +25,7 @@ type messageOutOfOrder struct {
 	explanation string
 }
 
-// MessageOutOfOrder returns a messagOutOfOrder PeerBehaviour.
+// MessageOutOfOrder returns a messageOutOfOrder PeerBehaviour.
 func MessageOutOfOrder(peerID p2p.ID, explanation string) PeerBehaviour {
 	return PeerBehaviour{peerID: peerID, reason: messageOutOfOrder{explanation}}
 }
