diff --git a/test/e2e/runner/exec.go b/test/e2e/runner/exec.go
index 6cccbf607..2a511182e 100644
--- a/test/e2e/runner/exec.go
+++ b/test/e2e/runner/exec.go
@@ -39,20 +39,20 @@ func execVerbose(args ...string) error {
 // execCompose runs a Docker Compose command for a testnet.
 func execCompose(dir string, args ...string) error {
 	return exec(append(
-		[]string{"docker-compose", "-f", filepath.Join(dir, "docker-compose.yml")},
+		[]string{"docker", "compose", "-f", filepath.Join(dir, "docker-compose.yml")},
 		args...)...)
 }
 
 func execComposeOutput(dir string, args ...string) ([]byte, error) {
 	return execOutput(append(
-		[]string{"docker-compose", "-f", filepath.Join(dir, "docker-compose.yml")},
+		[]string{"docker", "compose", "-f", filepath.Join(dir, "docker-compose.yml")},
 		args...)...)
 }
 
 // execComposeVerbose runs a Docker Compose command for a testnet and displays its output.
 func execComposeVerbose(dir string, args ...string) error {
 	return execVerbose(append(
-		[]string{"docker-compose", "-f", filepath.Join(dir, "docker-compose.yml")},
+		[]string{"docker", "compose", "-f", filepath.Join(dir, "docker-compose.yml")},
 		args...)...)
 }
 
