diff --git a/libs/protoio/writer.go b/libs/protoio/writer.go
index d4c66798f..f29065fc2 100644
--- a/libs/protoio/writer.go
+++ b/libs/protoio/writer.go
@@ -58,6 +58,7 @@ func (w *varintWriter) WriteMsg(msg proto.Message) (int, error) {
 			if n+binary.MaxVarintLen64 >= len(w.buffer) {
 				w.buffer = make([]byte, n+binary.MaxVarintLen64)
 			}
+			//nolint:gosec
 			lenOff := binary.PutUvarint(w.buffer, uint64(n))
 			_, err := m.MarshalTo(w.buffer[lenOff:])
 			if err != nil {
