diff --git a/rpc/core/abci.go b/rpc/core/abci.go
index ec0e67234..50ef3444c 100644
--- a/rpc/core/abci.go
+++ b/rpc/core/abci.go
@@ -17,7 +17,7 @@ func ABCIQuery(
 	height int64,
 	prove bool,
 ) (*ctypes.ResultABCIQuery, error) {
-	resQuery, err := env.ProxyAppQuery.QuerySync(abci.RequestQuery{
+	resQuery, err := GetEnvironment().ProxyAppQuery.QuerySync(abci.RequestQuery{
 		Path:   path,
 		Data:   data,
 		Height: height,
@@ -33,7 +33,7 @@ func ABCIQuery(
 // ABCIInfo gets some info about the application.
 // More: https://docs.cometbft.com/v0.34/rpc/#/ABCI/abci_info
 func ABCIInfo(ctx *rpctypes.Context) (*ctypes.ResultABCIInfo, error) {
-	resInfo, err := env.ProxyAppQuery.InfoSync(proxy.RequestInfo)
+	resInfo, err := GetEnvironment().ProxyAppQuery.InfoSync(proxy.RequestInfo)
 	if err != nil {
 		return nil, err
 	}
