diff --git a/rpc/core/evidence.go b/rpc/core/evidence.go
index bbfc1cbb5..bdb05293e 100644
--- a/rpc/core/evidence.go
+++ b/rpc/core/evidence.go
@@ -20,7 +20,7 @@ func BroadcastEvidence(ctx *rpctypes.Context, ev types.Evidence) (*ctypes.Result
 		return nil, fmt.Errorf("evidence.ValidateBasic failed: %w", err)
 	}
 
-	if err := env.EvidencePool.AddEvidence(ev); err != nil {
+	if err := GetEnvironment().EvidencePool.AddEvidence(ev); err != nil {
 		return nil, fmt.Errorf("failed to add evidence: %w", err)
 	}
 	return &ctypes.ResultBroadcastEvidence{Hash: ev.Hash()}, nil
