diff --git a/mempool/v0/doc.go b/mempool/v0/doc.go
index 3b5d0d20d..537bee9aa 100644
--- a/mempool/v0/doc.go
+++ b/mempool/v0/doc.go
@@ -15,7 +15,7 @@
 // 2. Mutations to the linked-list elements are atomic
 // 3. CheckTx() and/or ReapMaxBytesMaxGas() calls can be paused upon Update(), protected by .updateMtx
 
-// Garbage collection of old elements from mempool.txs is handlde via the
+// Garbage collection of old elements from mempool.txs is handled via the
 // DetachPrev() call, which makes old elements not reachable by peer
 // broadcastTxRoutine().
 
