diff --git a/cmd/cometbft/commands/compact.go b/cmd/cometbft/commands/compact.go
index 6febd9a1d..e1c22d511 100644
--- a/cmd/cometbft/commands/compact.go
+++ b/cmd/cometbft/commands/compact.go
@@ -18,8 +18,8 @@ var CompactGoLevelDBCmd = &cobra.Command{
 	Aliases: []string{"experimental_compact_goleveldb"},
 	Short:   "force compacts the CometBFT storage engine (only GoLevelDB supported)",
 	Long: `
-This is a temporary utility command that performs a force compaction on the state 
-and blockstores to reduce disk space for a pruning node. This should only be run 
+This is a temporary utility command that performs a force compaction on the state
+and blockstores to reduce disk space for a pruning node. This should only be run
 once the node has stopped. This command will likely be omitted in the future after
 the planned refactor to the storage engine.
 
