diff --git a/statesync/reactor.go b/statesync/reactor.go
index d8aa5b2a7..59f85f9aa 100644
--- a/statesync/reactor.go
+++ b/statesync/reactor.go
@@ -103,7 +103,7 @@ func (r *Reactor) RemovePeer(peer p2p.Peer, reason interface{}) {
 	}
 }
 
-// Receive implements p2p.Reactor.
+// ReceiveEnvelope implements p2p.Reactor.
 func (r *Reactor) ReceiveEnvelope(e p2p.Envelope) {
 	if !r.IsRunning() {
 		return
