diff --git a/statesync/snapshots.go b/statesync/snapshots.go
index efe705cc2..aac950bb5 100644
--- a/statesync/snapshots.go
+++ b/statesync/snapshots.go
@@ -21,7 +21,8 @@ type snapshot struct {
 	Hash     []byte
 	Metadata []byte
 
-	trustedAppHash []byte // populated by light client
+	trustedAppHash    []byte // populated by light client
+	trustedAppVersion uint64 // populated by light client
 }
 
 // Key generates a snapshot key, used for lookups. It takes into account not only the height and
