diff --git a/blockchain/v1/peer.go b/blockchain/v1/peer.go
index ad26585b3..8dbc8ee56 100644
--- a/blockchain/v1/peer.go
+++ b/blockchain/v1/peer.go
@@ -197,7 +197,7 @@ func BpPeerDefaultParams() *BpPeerParams {
 		timeout: 15 * time.Second,
 
 		// Minimum recv rate to ensure we're receiving blocks from a peer fast
-		// enough. If a peer is not sending data at at least that rate, we
+		// enough. If a peer is not sending data at least that rate, we
 		// consider them to have timedout and we disconnect.
 		//
 		// Assuming a DSL connection (not a good choice) 128 Kbps (upload) ~ 15 KB/s,
