diff --git a/mempool/mock/mempool.go b/mempool/mock/mempool.go
index 3f293381f..708471291 100644
--- a/mempool/mock/mempool.go
+++ b/mempool/mock/mempool.go
@@ -30,14 +30,15 @@ func (Mempool) Update(
 ) error {
 	return nil
 }
-func (Mempool) Flush()                        {}
-func (Mempool) FlushAppConn() error           { return nil }
-func (Mempool) TxsAvailable() <-chan struct{} { return make(chan struct{}) }
-func (Mempool) EnableTxsAvailable()           {}
-func (Mempool) SizeBytes() int64              { return 0 }
-
-func (Mempool) TxsFront() *clist.CElement    { return nil }
-func (Mempool) TxsWaitChan() <-chan struct{} { return nil }
+func (Mempool) Flush()                                    {}
+func (Mempool) FlushAppConn() error                       { return nil }
+func (Mempool) TxsAvailable() <-chan struct{}             { return make(chan struct{}) }
+func (Mempool) EnableTxsAvailable()                       {}
+func (Mempool) SizeBytes() int64                          { return 0 }
+func (m Mempool) GetTxByKey(types.TxKey) (types.Tx, bool) { return nil, false }
+func (m Mempool) WasRecentlyEvicted(types.TxKey) bool     { return false }
+func (Mempool) TxsFront() *clist.CElement                 { return nil }
+func (Mempool) TxsWaitChan() <-chan struct{}              { return nil }
 
 func (Mempool) InitWAL() error { return nil }
 func (Mempool) CloseWAL()      {}
