diff --git a/test/loadtime/cmd/load/main.go b/test/loadtime/cmd/load/main.go
index a0dba8ece..a90447ba9 100644
--- a/test/loadtime/cmd/load/main.go
+++ b/test/loadtime/cmd/load/main.go
@@ -3,8 +3,8 @@ package main
 import (
 	"fmt"
 
+	"github.com/cometbft/cometbft-load-test/pkg/loadtest"
 	"github.com/google/uuid"
-	"github.com/informalsystems/tm-load-test/pkg/loadtest"
 
 	"github.com/tendermint/tendermint/test/loadtime/payload"
 )
@@ -56,10 +56,13 @@ func (f *ClientFactory) ValidateConfig(cfg loadtest.Config) error {
 
 func (f *ClientFactory) NewClient(cfg loadtest.Config) (loadtest.Client, error) {
 	return &TxGenerator{
-		id:    f.ID,
+		id: f.ID,
+		//nolint:gosec
 		conns: uint64(cfg.Connections),
-		rate:  uint64(cfg.Rate),
-		size:  uint64(cfg.Size),
+		//nolint:gosec
+		rate: uint64(cfg.Rate),
+		//nolint:gosec
+		size: uint64(cfg.Size),
 	}, nil
 }
 
