diff --git a/types/event_bus.go b/types/event_bus.go
index 27aff09f1..22ffe79ec 100644
--- a/types/event_bus.go
+++ b/types/event_bus.go
@@ -144,6 +144,10 @@ func (b *EventBus) PublishEventNewBlock(data EventDataNewBlock) error {
 	return b.pubsub.PublishWithEvents(ctx, data, events)
 }
 
+func (b *EventBus) PublishEventNewSignedBlock(data EventDataSignedBlock) error {
+	return b.Publish(EventSignedBlock, data)
+}
+
 func (b *EventBus) PublishEventNewBlockHeader(data EventDataNewBlockHeader) error {
 	// no explicit deadline for publishing events
 	ctx := context.Background()
@@ -251,6 +255,10 @@ func (NopEventBus) PublishEventNewBlock(data EventDataNewBlock) error {
 	return nil
 }
 
+func (NopEventBus) PublishEventNewSignedBlock(data EventDataSignedBlock) error {
+	return nil
+}
+
 func (NopEventBus) PublishEventNewBlockHeader(data EventDataNewBlockHeader) error {
 	return nil
 }
